### PR TITLE
Chad/#103 transition minutes from api to library

### PR DESCRIPTION
**Pull request by chadlagore**
Handles #103 #108 #111 

## :construction_worker: Changes

+ Replaced Docker app with a library!
+ Reused most of old code (except YouTube stuff).
+ **Breaking change**: replaced `matplotlib` spectrograms with `scipy.signal.spectrogram` which produces a flat (2d spectrogram). This will likely reduce training time at (hopefully) no accuracy cost.
+ Amazingly, the second of these two spectrogram types performs (much) better 🤔 
image
image
+ Added travis build and coveralls token.
+ @iKevinY added `pipenv`!!! :tada:

### Typical Usage

```python
from minutes import Speaker, Minutes

minutes = Minutes(ms_per_observation=500, model='cnn')

# Create some speakers with some audio.
speaker1 = Speaker('speaker1')
speaker1.add_audio('path/to/audio1.wav')

speaker2 = Speaker('speaker2')
speaker2.add_audio('path/to/audio2.wav')

# Add speakers to the model.
minutes.add_speakers([speaker1, speaker2])

# Fit the model.
minutes.fit()  # Currently breaks (have to refit base model).
result = minutes.predict()
```

```python
from minutes import Speaker, Minutes
from minutes.base import BaseModel

model = BaseModel('my_cnn_base', ms_per_observation=500)

# Create some speakers with some (large) audio.
speaker1 = Speaker('speaker1')
speaker1.add_audio('path/to/audio1.wav')

speaker2 = Speaker('speaker2')
speaker2.add_audio('path/to/audio2.wav')

# Add speakers to the model.
minutes.add_speakers([speaker1, speaker2])

# Fit the model.
model.fit()  # Prints validation results....
model.save()

# Use the new base model.
minutes = Minutes(model='my_cnn_base')
# ... add speakers, predict etc.
```

## :flashlight: Testing Instructions

For now,

```
$ py.test -vvv --cov=minutes test
```

Lets keep the coverage up!

## Repo Layout

```
.
├── README.md
├── bld.bat
├── build.sh
├── environment.yml
├── meta.yaml
├── minutes
│   ├── __init__.py
│   ├── audio.py
│   ├── base.py
│   ├── conversation.py
│   ├── minutes.py
│   ├── models
│   │   ├── __init__.py
│   │   └── cnn.h5
│   └── speaker.py
├── setup.py
└── test
    ├── __init__.py
    ├── config.py
    ├── fixtures
    │   ├── sample1.wav
    │   └── sample2.wav
    ├── test_audio.py
    ├── test_base.py
    ├── test_minutes.py
    └── test_speaker.py
```

## Base Model(s)

- I retrained `minutes/models/cnn.5` on flat spectrograms (no channels).
- The code to train this is in the `BaseModel.fit` method.
- The model is much simpler than our previous one (ready for optimization!) but still achieves >90% validation accuracy on 3 speakers after 50 epochs.
- Note: I had a model without `Dropout`, it achieved > 95% validation accuracy, but performed far worse in a transfer situation. This makes sense, considering dropout reduces overfitting to the current speakers!

```python
base_model = BaseModel('taco', ms_per_observation=3000)
speaker1 = Speaker('4640')
speaker2 = Speaker('8098')
speaker3 = Speaker('441')

speaker1.add_audio('test/fixtures/4640')
speaker2.add_audio('test/fixtures/8098')
speaker3.add_audio('test/fixtures/441')

base_model.add_speaker(speaker1)
base_model.add_speaker(speaker2)
base_model.add_speaker(speaker3)

base_model.fit(verbose=2)
```
```
...
Epoch 45/50
 - 3s - loss: 0.6708 - acc: 0.8345 - val_loss: 0.6273 - val_acc: 0.9076
Epoch 46/50
 - 3s - loss: 0.6404 - acc: 0.8414 - val_loss: 0.6115 - val_acc: 0.9076
Epoch 47/50
 - 3s - loss: 0.6175 - acc: 0.8454 - val_loss: 0.5964 - val_acc: 0.9076
Epoch 48/50
 - 3s - loss: 0.5998 - acc: 0.8553 - val_loss: 0.5815 - val_acc: 0.9076
Epoch 49/50
 - 3s - loss: 0.6101 - acc: 0.8583 - val_loss: 0.5678 - val_acc: 0.9056
Epoch 50/50
 - 3s - loss: 0.5849 - acc: 0.8632 - val_loss: 0.5533 - val_acc: 0.9137
```
```python
base_model.model.save('minutes/models/cnn.h5')
base_model.model.summary()
```
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv1d_3 (Conv1D)            (None, 49, 32)            219168    
_________________________________________________________________
max_pooling1d_3 (MaxPooling1 (None, 24, 32)            0         
_________________________________________________________________
dropout_3 (Dropout)          (None, 24, 32)            0         
_________________________________________________________________
flatten_3 (Flatten)          (None, 768)               0         
_________________________________________________________________
dense_21 (Dense)             (None, 128)               98432     
_________________________________________________________________
dense_22 (Dense)             (None, 3)                 387       
=================================================================
Total params: 317,987
Trainable params: 317,987
Non-trainable params: 0
_________________________________________________________________
```

## Flac-Wav Conversion

Rather crudely, I did something like:

```python
import glob
from subprocess import call

flac_files = glob.glob('test/fixtures/5561' + '/**/*.flac', recursive=True)

for file in flac_files:
    call(["ffmpeg", "-i", file, file.strip('.flac') + '.wav'])
```

This is obviously not an option for the library. We should find a way to read in `.flac` files properly (#110).

## Other Notes

- ~3000ms observations gave me the best accuracy.
- Transfer accuracy results:

```
Epoch 12/15
 - 1s - loss: 0.7596 - acc: 0.8963 - val_loss: 0.7163 - val_acc: 0.9102
Epoch 13/15
 - 1s - loss: 0.7092 - acc: 0.8872 - val_loss: 0.6522 - val_acc: 0.9224
Epoch 14/15
 - 1s - loss: 0.6473 - acc: 0.8943 - val_loss: 0.5936 - val_acc: 0.9163
Epoch 15/15
 - 2s - loss: 0.5940 - acc: 0.9023 - val_loss: 0.5443 - val_acc: 0.9286
```

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,41 @@
 sudo: required
+dist: trusty
+language: python
+matrix:
+    include:
+        - python: 2.7
```
Comment (iKevinY) [inline .travis.yml]: I don't think that tests are actually running on Python 2.7 — for the most recent job, the `pytest` runner is actually [executing on Python 3.6](https://travis-ci.org/ubclaunchpad/minutes/jobs/378193136#L813) (there are a few instances of f-strings which won't work on 2.7!)

File: test/test_minutes.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from minutes import Minutes
+import test.config as c
+
+
+def test_train():
+    minutes = Minutes(ms_per_observation=100)
```
Comment (iKevinY) [inline test/test_minutes.py]: It might be good to go through and test model training with all of the models declared in `Minutes.models`.

File: README.md [hunk 3]
```diff
@@ -4,9 +4,13 @@
 
 [![Build Status](https://travis-ci.org/ubclaunchpad/minutes.svg?branch=master)](https://travis-ci.org/ubclaunchpad/minutes)
 
-Audio speaker diarization and transcription API.
+[![Coverage Status](https://coveralls.io/repos/github/ubclaunchpad/minutes/badge.svg)](https://coveralls.io/github/ubclaunchpad/minutes)
 
-## :running: Getting Started
+Audio speaker diarization and transcription library.
```
Comment (iKevinY) [inline README.md]: Are we going to ship transcription with Minutes as well? It'll complicate the library a bit, since we'd probably want to call out to Google Cloud Speech-to-Text (unless there's a nice offline library that does it for us).

File: README.md [hunk 4]
```diff
@@ -15,68 +19,38 @@ conda env create -f environment.yml
...
-
-```
-make push-dev
-make push-prod
+```bash
+conda-build .
 ```
 
-You will need to set `DOCKER_USERNAME` and `DOCKER_PASSWORD` and be a member of the `ubclaunchpad` docker hub organization to deploy manually.
+## Example Usage
```
Comment (iKevinY) [inline README.md]: It might also be useful to provide an example of a call to `Minutes.models` that returns a list of available models, to showcase what base models come prepackaged with the library.

File: meta.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  run:
+    - python
+
+test:
+  imports:
+    - minutes
+
+about:
+  home: https://github.com/ubclaunchpad/minutes
+  license: BSD
```
Comment (iKevinY) [inline meta.yaml]: Not MIT? 😁

File: .travis.yml (hunk 1 above)
Comment (chadlagore) [reply .travis.yml]: Its true, we totally don't support 2.7 😬, everything on travis is run through a `conda` environment.

I typically let it run on both interpreters in travis in case someone bootstraps the project with their local Python, before using an env. But at the moment, no environment-less invocation of `python` exists, so it's a meaningless run.

File: README.md (hunk 3 above)
Comment (chadlagore) [reply README.md]: Good call, I'll remove this.

File: minutes/audio.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            # Collect all data and concatenate it.
+            for file in wav_files:
+                data, rate = sf.read(file)
+                audio += data,
+                rates += rate,
+
+            self.data = np.concatenate(audio)
+
+            # Set rate as mode of rates for now.
+            # TODO: Resample audio?
```
Comment (chadlagore) [inline minutes/audio.py]: https://github.com/ubclaunchpad/minutes/issues/113

File: minutes/audio.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        data = data[:N * d]
+
+        if verbose:
+            t = len(self.data) - (N * d)
+            print('Truncating {} bytes from end of sample'.format(t))
+
+        # Reshape for processing into spectrograms.
+        data = data.reshape((N, d))
+
+        def spec_from_row(row):
```
Comment (chadlagore) [inline minutes/audio.py]: https://github.com/ubclaunchpad/minutes/issues/114

File: minutes/base.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        obs = [s.get_observations(self.ms_per_observation)
+               for s in self.speakers]
+        labels = [[i] * len(o) for i, o in enumerate(obs)]
+        flattened_labels = [j for i in labels for j in i]
+
+        y = to_categorical(flattened_labels)
+        X = np.concatenate(obs)
+
+        # TODO: Propagate more configuration options to user.
```
Comment (chadlagore) [inline minutes/base.py]: https://github.com/ubclaunchpad/minutes/issues/115

**Comment by iKevinY** [comment]
What code is generating the two spectrograms that are in your PR description? I'm curious why so much of the second one is purple (what information is it actually encoding, compared to the first?) 😮

File: minutes/base.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+        self.model.fit(
+            X_train, y_train,
+            validation_data=(X_test, y_test),
+            epochs=50, batch_size=16, verbose=verbose
+        )
+
+    def save(self):
+        """Save the model... somewhere."""
+        self.model.save(os.path.join(MODELS_DIR, self.name + '.h5'))
```
Comment (chadlagore) [inline minutes/base.py]: https://github.com/ubclaunchpad/minutes/issues/116

**Comment by chadlagore** [comment]
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.spectrogram.html `mode="phase"` produces the first spectrogram. I agree that the fact that the purple spectrogram learns is rather surprising. #114 will allow users to configure this a bit more.
